### PR TITLE
Added mouse handlers to GeoJSONLayer

### PR DESCRIPTION
**Pull request by SBRK**
I needed a way to know when mouse was hovering/clicking on GeoJSONLayer features

### CODE REVIEW

**Comment by alex3165** [comment]
Cool, this PR should fix https://github.com/alex3165/react-mapbox-gl/issues/325, I will take a look at it this evening 👍

File: src/geojson-layer.ts [hunk 1]
```diff
@@ -98,8 +107,26 @@ export default class GeoJSONLayer extends React.Component<Props, {}> {
...
       before
     );
+
+    this.mapLayerMouseHandlers(type)
   };
 
+  private mapLayerMouseHandlers = (type: string) => {
+    const { map } = this.context;
+
+    const layerId = `${this.id}-${type}`;
```
Comment (alex3165) [inline src/geojson-layer.ts]: Could you create a function `buildLayerId` and use it in line 88 too?

Comment (SBRK) [reply src/geojson-layer.ts]: done

**Comment by alex3165** [comment]
Hey @SBRK , I had a look at the code and it look good. Could listen off the event on `componentUnmount` ? Probably something like:
```
this.layerIds.forEach(id => {
  Object.keys(eventToHandler).forEach(event => {
    const handler = this.props[`${typeToLayerLUT[type]}${eventToHandler[event]}`] || null;
    if (handler) {
      map.off(event, id);
    }
  });
  map.removeLayer(lId)
});
```

**Comment by SBRK** [comment]
Actually, managing the `off` might be a bit trickier, as it requires to pass the handler. I would also need to manage changing listeners in the `componentWillReceiveProps`, I didn't put time into it yet, as it's not a case I needed to handle for the moment. Will do it during the day if I have time (as well as documentation)

**Comment by SBRK** [comment]
@alex3165 I added the `componentWillReceiveProps` / `componentWillUnmount` code as well as the doc

File: src/geojson-layer.ts [hunk 2]
```diff
@@ -137,6 +168,20 @@ export default class GeoJSONLayer extends React.Component<Props, {}> {
     map.removeSource(this.id);
     map.off('styledata', this.onStyleDataChange);
     this.layerIds.forEach(lId => map.removeLayer(lId));
+
+    for (const type in typeToLayerLUT) {
```
Comment (alex3165) [inline src/geojson-layer.ts]: Could you use `Object.keys(typeToLayerLUT).forEach ` to keep the coding style consistent with the rest of the codebase? Same for eventToHandler and same for lines 223 to 227

File: src/geojson-layer.ts [hunk 3]
```diff
@@ -169,11 +214,31 @@ export default class GeoJSONLayer extends React.Component<Props, {}> {
...
 
           Object.keys(layoutDiff).forEach(key => {
-            map.setLayoutProperty(`${this.id}-${type}`, key, layoutDiff[key]);
+            map.setLayoutProperty(this.buildLayerId(type), key, layoutDiff[key]);
           });
         }
       }
     }
+
+    for (const type in typeToLayerLUT) {
```
Comment (alex3165) [inline src/geojson-layer.ts]: Line 209 we already iterate over the `typeToLayerLUT`, if you could insert your block in there that would be better. Also it is a technical debt I would rather use `Object.keys().forEach` too

File: src/geojson-layer.ts [hunk 4]
```diff
@@ -169,11 +214,31 @@ export default class GeoJSONLayer extends React.Component<Props, {}> {
...
+      if (typeToLayerLUT.hasOwnProperty(type)) {
+        for (const event in eventToHandler) {
+          if (eventToHandler.hasOwnProperty(event)) {
+            const prop = typeToLayerLUT[type] + eventToHandler[event];
+
+            if (!isEqual(props[prop], this.props[prop])) {
```
Comment (alex3165) [inline src/geojson-layer.ts]: To test equality on a function I think you can go for a reference check straight away so `props[prop] === this.props[prop]` no need for value equality and `===` is much faster

**Comment by SBRK** [comment]
@alex3165 all done. I regrouped the change of layout/paint/handlers in a same `Object.keys(typeToLayerLUT).forEach`

**Comment by alex3165** [comment]
Alright, it is looking good, there one more thing missing, we need to add all the events to the typescript interface, I think it would be better to refactor a small bit too:

Could you extract all the props per shape type from the global Props interface (example for line):
```
export interface LineProps {
  lineLayout?: MapboxGL.LineLayout;
  linePaint?: MapboxGL.LinePaint;
  lineOnMouseDown?: MouseEvent;
  lineOnMouseUp?: MouseEvent;
  lineOnMouseMove?: MouseEvent;
  lineOnMouseEnter?: MouseEvent;
  lineOnMouseLeave?: MouseEvent;
  lineOnClick?: MouseEvent;
}
```

Where MapEvent is:
```
export type MouseEvent = (evt: Event) => any
```

In the end you can compose everything back to the global `Props` by doing:
```
export interface Props extends LineProps, ... {
...
}
```

If you don't feel confortable with those changes I can add them once we merge this PR it is fine

**Comment by SBRK** [comment]
I'm not familiar with TypeScript so I think it's better if you do it :)

**Comment by alex3165** [comment]
Alright, it look good to me I will add the missing stuff (TS, Tests, Example) after the merge, everything good for you @SBRK ?

**Comment by alex3165** [comment]
For history, this PR should also solve https://github.com/alex3165/react-mapbox-gl/issues/312

**Pull request merged**